numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -357,11 +358,24 @@ array_implement_array_function(
             }
             Py_DECREF(tmp_has_override);
             PyDict_DelItem(kwargs, npy_ma_str_like);
+            like_removed = 1;
         }
     }
 
-    PyObject *res = array_implement_array_function_internal(
-        public_api, relevant_args, args, kwargs);
+    /*
+     * If `like=` kwarg was removed, `implementation` points to the NumPy
+     * public API, as `public_api` is in that case the wrapper dispatcher
+     * function. For example, in the `np.full` case, `implementation` is
+     * `np.full`, whereas `public_api` is `_full_with_like`. This is done
+     * to ensure `__array_function__` implementations can do equality/identity
+     * comparisons when `like=` is present.
+     */
+    if (like_removed)
+        res = array_implement_array_function_internal(
+            implementation, relevant_args, args, kwargs);
```

Comment:
Could you add braces, we always put them in NumPy.

This is a surprisingly small change!  Could also use `public_api = implementation` with a comment above, but this is OK with me.  I do wonder a bit if it might not be nicer to handle this differently, but so long we have the special case here to drop `like`, I agree that this is completely fine.

Thanks!